mountpoint-s3/src/inode.rs

```diff
@@ -130,7 +130,20 @@ impl Superblock {
                 if new_lookup_count == 0 {
                     // Safe to remove, kernel no longer has a reference to it.
                     trace!(ino, "removing inode from superblock");
-                    inodes.remove(&ino);
+                    let inode = inodes.remove(&ino).expect("we just retrieved it");
+
+                    // Should be impossible for this to fail (VFS inodes reference their parent, so
+                    // children need to be freed first), but let's not crash in a `forget` function...
+                    let Some(parent) = inodes.get(&inode.parent()) else {
+                        debug_assert!(false, "children should be forgotten before parents");
+                        return;
+                    };
+                    let mut parent_state = parent.inner.sync.write().unwrap();
```

Comment:
We typically use `parent.get_mut_inode_state()`, unless we also want directories marked as deleted?

Reply:
We do want those directories here, otherwise we'd leak them, since `forget` is the only way to free them.